File: scala/space-age/src/main/scala/SpaceAge.scala
case class SpaceAge(seconds: Long) {

  val SecondsInEarthYear = 31557600.0

  val EarthOrbitalPeriodInEarthYears = 1.0
  val MercuryOrbitalPeriodInEarthYears = 0.2408467
  val VenusOrbitalPeriodInEarthYears = 0.61519726
  val MarsOrbitalPeriodInEarthYears = 1.8808158
  val JupiterOrbitalPeriodInEarthYears = 11.862615
  val SaturnOrbitalPeriodInEarthYears = 29.447498
  val UranusOrbitalPeriodInEarthYears = 84.016846
  val NeptureOrbitalPeriodInEarthYears = 164.79132

  def round(number: Double, digits: Int) = Math.round(number * Math.pow(10, digits)) / Math.pow(10, digits)

  def getYears(orbitalPeriod: Double): Double = round(seconds.toDouble / (SecondsInEarthYear * orbitalPeriod), 2)

  def onEarth(): Double = getYears(EarthOrbitalPeriodInEarthYears)
  def onMercury(): Double = getYears(MercuryOrbitalPeriodInEarthYears)
  def onVenus(): Double = getYears(VenusOrbitalPeriodInEarthYears)
  def onMars(): Double = getYears(MarsOrbitalPeriodInEarthYears)
  def onJupiter(): Double = getYears(JupiterOrbitalPeriodInEarthYears)
  def onSaturn(): Double = getYears(SaturnOrbitalPeriodInEarthYears)
  def onUranus(): Double = getYears(UranusOrbitalPeriodInEarthYears)
  def onNeptune(): Double = getYears(NeptureOrbitalPeriodInEarthYears)
}

File: typescript/space-age/space-age.ts
export default class SpaceAge {
  public seconds: number

  constructor(seconds: number) {
    this.seconds = seconds
  }

  public onMercury(): number {
    return this.calculateAge(0.2408467)
  }

  public onVenus(): number {
    return this.calculateAge(0.61519726)
  }

  public onEarth(): number {
    return this.calculateAge(1)
  }

  public onMars(): number {
    return this.calculateAge(1.8808158)
  }

  public onJupiter(): number {
    return this.calculateAge(11.862615)
  }

  public onSaturn() {
    return this.calculateAge(29.447498)
  }

  public onUranus(): number {
    return this.calculateAge(84.016846)
  }

  public onNeptune(): number {
    return this.calculateAge(164.79132)
  }

  private calculateAge(periodInEarthYears: number): number {
    const earthOrbitInSeconds = 31557600
    return parseFloat(
      (this.seconds / (earthOrbitInSeconds * periodInEarthYears)).toFixed(2)
    )
  }
}
